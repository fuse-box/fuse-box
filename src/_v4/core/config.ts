import * as appRoot from "app-root-path";
import { ensureAbsolutePath } from "../utils/utils";
import { IConfig } from "./interfaces";

export function createConfig(props: IConfig): IConfig {
	const config: IConfig = {
		root: process.env.APP_ROOT || appRoot.path,
	};

	config.defaultCollectionName = "default";

	if (props.homeDir) {
		config.homeDir = ensureAbsolutePath(props.homeDir, config.root);
	}

	if (props.modules) {
		config.modules = props.modules;
	}

	if (props.output) {
		config.output = props.output;
	}

	if (props.target) {
		config.target = props.target;
	}

	if (props.fuseBoxPolyfillsFolder) {
		config.fuseBoxPolyfillsFolder = props.fuseBoxPolyfillsFolder;
	}
	return config;
}
