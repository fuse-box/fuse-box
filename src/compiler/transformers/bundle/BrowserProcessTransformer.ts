import { IVisit, IVisitorMod } from '../../Visitor/Visitor';
import { ITransformerSharedOptions } from '../../interfaces/ITransformerSharedOptions';
import {
  isPropertyOrPropertyAccess,
  createASTFromObject,
  defineVariable,
  createRequireStatement,
} from '../../Visitor/helpers';
import { ImportType } from '../../interfaces/ImportType';

interface BrowserProcessTransformProps {
  env?: { [key: string]: string };
}
export type IBrowserProcessTransform = BrowserProcessTransformProps & ITransformerSharedOptions;
export function BrowserProcessTransformer(options: IBrowserProcessTransform) {
  options = options || {};

  let globalEnvInserted = false;
  let entireEnvInserted = false;
  return (visit: IVisit): IVisitorMod => {
    const { node, parent } = visit;

    const accessList = isPropertyOrPropertyAccess(node, parent, 'process');
    if (accessList) {
      const variableAmount = accessList.length;
      if (variableAmount === 3 && accessList[1] === 'env') {
        const keyName = accessList[2];

        if (options.env && options.env[keyName] !== undefined) {
          return { replaceWith: { type: 'Literal', value: options.env[keyName].toString() } };
        }
        return { replaceWith: { type: 'Identifier', value: 'undefined' } };
      }
      if (variableAmount === 2) {
        switch (accessList[1]) {
          case 'version':
            return { replaceWith: { type: 'Literal', value: process.version }, avoidReVisit: true };
          case 'versions':
            return { replaceWith: { type: 'ObjectExpression', properties: [] }, avoidReVisit: true };
          case 'title':
            return { replaceWith: { type: 'Literal', value: 'browser' }, avoidReVisit: true };
          case 'umask':
            return { replaceWith: { type: 'Literal', value: 0 }, avoidReVisit: true };
          case 'browser':
            return { replaceWith: { type: 'Literal', value: true }, avoidReVisit: true };
          case 'cwd':
            return { replaceWith: { type: 'Literal', value: './' }, avoidReVisit: true };
          case 'env':
            if (options.env) {
              const response: IVisitorMod = {};
              if (!globalEnvInserted) {
                response.prependToBody = [defineVariable('___env', createASTFromObject(options.env))];
                globalEnvInserted = true;
              }
              response.replaceWith = { type: 'Identifier', name: '___env' };
              return response;
            } else return { replaceWith: createASTFromObject({}), avoidReVisit: true };

          default:
            // inserting require("process")
            // because we cannot match a know variable
            if (!entireEnvInserted) {
              entireEnvInserted = true;
              const statement = createRequireStatement('process', 'process');
              if (options.onRequireCallExpression)
                options.onRequireCallExpression(ImportType.REQUIRE, statement.reqStatement);
              return {
                avoidReVisit: true,
                prependToBody: [statement.statement],
              };
            }
            break;
        }
      }
    }
    return;
  };
}
