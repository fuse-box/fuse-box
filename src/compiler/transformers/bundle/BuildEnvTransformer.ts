import { IVisit, IVisitorMod } from '../../Visitor/Visitor';
import { ASTNode, ASTType } from '../../interfaces/AST';
import { ITransformer } from '../../interfaces/ITransformer';
import { ITransformerSharedOptions } from '../../interfaces/ITransformerSharedOptions';
import { parseJavascript } from '../../parser';

interface BrowserProcessTransformProps {
  env?: { [key: string]: string };
}
export type IBrowserProcessTransform = BrowserProcessTransformProps & ITransformerSharedOptions;

export const BUILD_ENV_NAME = '__build_env';

const CACHE_AST: Record<string, ASTNode> = {};

function getEnvAST(obj: any) {
  let str;
  if (typeof obj === 'string') str = obj;
  else str = JSON.stringify(obj);
  if (CACHE_AST[str]) return CACHE_AST[str];

  let ast;

  if (typeof obj === 'object') {
    ast = parseJavascript('var a = ' + str);
  } else {
    try {
      ast = parseJavascript(str);
    } catch (e) {
      throw new Error(`Error while parsing string : "${str}".
Make sure you're passing a valid javascript string.
Otherwise pass a javascript object (it will be converting to AST automatically)
`);
    }
  }
  const body = ast.body as Array<ASTNode>;
  const firstItem = body[0];
  if (firstItem) {
    let expression;
    switch (firstItem.type) {
      case ASTType.ExpressionStatement:
        expression = firstItem.expression;
        break;
      case ASTType.VariableDeclaration:
        expression = firstItem.declarations[0].init;
        break;
      default:
        break;
    }
    if (expression) CACHE_AST[str] = expression;
    return expression;
  }
  return CACHE_AST[str];
}

export function BuildEnvTransformer(): ITransformer {
  return {
    commonVisitors: props => {
      const buildEnv = props.compilerOptions.buildEnv;
      return {
        onEachNode: (visit: IVisit): IVisitorMod => {
          const { node } = visit;
          if (node.type === ASTType.MemberExpression && node.object.name === BUILD_ENV_NAME) {
            const propertyName = node.property.name;
            if (buildEnv[propertyName] !== undefined) {
              const ast = getEnvAST(buildEnv[propertyName]);
              if (ast) return { replaceWith: ast };
            }
            return { replaceWith: { name: 'undefined', type: ASTType.Identifier } as ASTNode };
          }
          return;
        },
      };
    },
  };
}
