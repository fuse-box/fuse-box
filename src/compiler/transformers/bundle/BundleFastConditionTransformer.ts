import { IVisit, IVisitorMod } from '../../Visitor/Visitor';
import { ASTNode } from '../../interfaces/AST';
import { ITransformer } from '../../interfaces/ITransformer';
import { computeBoolean } from '../../static_compute/computeBoolean';

export interface IBundleFastConditionTransformer {
  isBrowser?: boolean;
  isServer?: boolean;
  env: { [key: string]: any };
}

export function BundleFastConditionUnwrapper(): ITransformer {
  return {
    commonVisitors: props => {
      const compilerOptions = props.compilerOptions;
      const env = compilerOptions.processEnv;
      const isBrowser = compilerOptions.buildTarget === 'browser';
      const isServer = compilerOptions.buildTarget === 'server';

      return {
        onEachNode: (visit: IVisit): IVisitorMod => {
          const { node } = visit;
          if (node.type === 'IfStatement') {
            let operator = node.test.operator;
            let rightValue;
            let target: ASTNode;
            if (node.test) {
              if (node.test.type === 'MemberExpression') {
                target = node.test;
              }
              const left = node.test.left;
              const right = node.test.right;
              if (right && right.type === 'Literal') {
                rightValue = right.value;
              }
              if (left) {
                if (left.type == 'MemberExpression') target = left;
              }
            }
            let replacement: ASTNode;
            let result: boolean;

            let shouldTranspile = false;

            if (target) {
              // process.env.*
              if (
                target.object.type === 'MemberExpression' &&
                target.object.object.name === 'process' &&
                target.object.property.name === 'env'
              ) {
                if (rightValue && operator) {
                  shouldTranspile = true;
                  result = computeBoolean(env[target.property.name], operator, rightValue);
                  if (result) replacement = node.consequent;
                  else replacement = node.alternate;
                }
              }
              if (target.object.name === 'FuseBox') {
                if (target.property.name === 'isBrowser') {
                  shouldTranspile = true;
                  if (isBrowser) replacement = node.consequent;
                  else replacement = node.alternate;
                }

                if (target.property.name === 'isServer') {
                  shouldTranspile = true;
                  if (isServer) replacement = node.consequent;
                  else replacement = node.alternate;
                }
              }
            }

            if (shouldTranspile) {
              if (replacement) {
                if (replacement.body) return { replaceWith: replacement.body };
                else if (replacement.type === 'ExpressionStatement') return { replaceWith: replacement };
              } else {
                return { removeNode: true };
              }
            }
          }
          return;
        },
      };
    },
  };
}
