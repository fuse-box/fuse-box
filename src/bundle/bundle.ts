import * as CleanCSS from 'clean-css';
import * as convertSourceMap from 'convert-source-map';
import * as offsetLinesModule from 'offset-sourcemap-lines';
import * as path from 'path';
import * as Terser from 'terser';
import { BundleSource, createBundleSource } from '../bundleRuntime/bundleSource';
import { Context } from '../core/context';
import { IModule } from '../moduleResolver/module';
import { IOutputBundleConfigAdvanced } from '../output/OutputConfigInterface';
import { distWriter, IWriterConfig } from '../output/distWriter';
import { Concat } from '../utils/utils';

export interface Bundle {
  config: IWriterConfig;
  containsAPI?: boolean;
  containsApplicationEntryCall?: boolean;
  contents: string;
  data: Concat;
  entries?: Array<IModule>;
  isCSSType?: boolean;
  priority: number;
  source: BundleSource;
  type: BundleType;
  webIndexed: boolean;
  createSourceMap: (sourceMap: string) => Promise<void>;
  generate: (opts?: { runtimeCore?: string }) => Promise<IBundleWriteResponse>;
  generateHMRUpdate?: () => string;
  prepare: () => IWriterConfig;
  write: () => Promise<IBundleWriteResponse>;
}

export enum BundleType {
  CSS_APP = 1,
  CSS_SPLIT = 2,
  JS_APP = 3,
  JS_SERVER_ENTRY = 4,
  JS_SPLIT = 5,
  JS_VENDOR = 6,
}

export interface IBundleProps {
  bundleConfig?: IOutputBundleConfigAdvanced;
  ctx: Context;
  fileName?: string;
  priority?: number;
  type?: BundleType;
  webIndexed?: boolean;
}

export interface IBundleWriteResponse {
  absPath: string;
  browserPath: string;
  bundle?: Bundle;
  relativePath: string;
}

export function createBundle(props: IBundleProps): Bundle {
  const { bundleConfig, ctx, priority, type, webIndexed = true } = props;
  const outputConfig = ctx.outputConfig;
  const isProduction = ctx.config.isProduction;
  const target = ctx.config.target;

  const bundleWriter = distWriter({ hashEnabled: isProduction, root: outputConfig.distRoot });

  const isCSS = type === BundleType.CSS_APP || type === BundleType.CSS_SPLIT;
  const source = createBundleSource({
    isCSS: isCSS,
    isProduction: props.ctx.config.isProduction,
    target,
  });

  const shouldCleanCSS = !!ctx.config.cleanCSS;
  function optimizeCSS(self: Bundle) {
    let userProps = {};
    if (typeof ctx.config.cleanCSS === 'object') userProps = ctx.config.cleanCSS;

    const response = new CleanCSS({
      ...userProps,
      sourceMap: source.containsMaps,
      sourceMapInlineSources: true,
    }).minify(self.data.content.toString(), self.data.sourceMap);
    self.data = ({ sourceMap: response.sourceMap, content: response.styles } as unknown) as Concat;
  }

  const self: Bundle = {
    config: null,
    contents: null,
    data: null,
    isCSSType: isCSS,
    priority,
    source,
    type,
    webIndexed,
    createSourceMap: async (sourceMap: string) => {
      const sourceMapName = path.basename(self.config.relativePath) + '.map';
      if (isCSS) {
        // just in case remove the existing sourcemap references
        // some css preprocessors add it
        self.contents = self.contents.replace(/\/\*\#\s?sourceMappingURL.*?\*\//g, '');
        self.contents += `\n/*#  sourceMappingURL=${sourceMapName} */`;
      } else {
        self.contents += `\n//# sourceMappingURL=${sourceMapName}`;
      }
      const targetDir = path.dirname(self.config.absPath);
      const sourceMapFile = path.join(targetDir, sourceMapName);

      await bundleWriter.write(sourceMapFile, sourceMap);
    },
    generate: async (opts?: { runtimeCore?: string; uglify?: boolean }) => {
      opts = opts || {};
      if (!self.config) self.prepare();
      if (self.entries) source.entries = self.entries;

      ctx.ict.sync('before_bundle_write', { bundle: self });
      self.data = source.generate({ isIsolated: bundleConfig.isolatedApi, runtimeCore: opts.runtimeCore });

      if (isCSS && shouldCleanCSS) optimizeCSS(self);

      self.contents = self.data.content.toString();

      let sourceMap;
      if (source.containsMaps && self.data.sourceMap) {
        sourceMap = self.data.sourceMap.toString();
      }

      if (ctx.config.isProduction && !self.isCSSType && opts.uglify) {
        const terserOpts: any = {
          sourceMap: source.containsMaps
            ? {
                content: self.data.sourceMap,
                includeSources: true,
              }
            : undefined,
        };
        ctx.log.info('minify', self.config.absPath);
        const result = Terser.minify(self.contents, terserOpts);
        self.contents = result.code;
        if (source.containsMaps && result.map) {
          sourceMap = result.map.toString();
        }
      }
      // writing source maps
      if (source.containsMaps) await self.createSourceMap(sourceMap);
      // write the bundle to fs
      return await self.write();
    },
    generateHMRUpdate: (): string => {
      const concat = source.generate({ isIsolated: false, runtimeCore: undefined });
      const rawSourceMap = concat.sourceMap;
      let stringContent = concat.content.toString();
      if (self.source.containsMaps) {
        if (rawSourceMap) {
          let json = JSON.parse(rawSourceMap);
          // since new Function wrapoer adds extra 2 lines we need to shift sourcemaps
          json = offsetLinesModule(json, 2);
          const sm = convertSourceMap.fromObject(json).toComment();
          stringContent += '\n' + sm;
        }
      }
      return stringContent;
    },
    prepare: (): IWriterConfig => {
      self.config = bundleWriter.createWriter({
        fileName: props.fileName,
        hash: isProduction && self.source.generateHash(),
        publicPath: bundleConfig.publicPath,
        userString: bundleConfig.path,
      });
      return self.config;
    },
    write: async () => {
      await bundleWriter.write(self.config.absPath, self.contents);
      return {
        absPath: self.config.absPath,
        browserPath: self.config.browserPath,
        bundle: self,
        relativePath: self.config.relativePath,
      };
    },
  };
  return self;
}
