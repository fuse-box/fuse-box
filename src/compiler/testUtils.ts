import { ICompilerOptions } from '../compilerOptions/interfaces';
import { transformModule } from './core/transformModule';
import { generate } from './generator/generator';
import { ASTNode } from './interfaces/AST';
import { ITransformer, ITransformerCommon } from './interfaces/ITransformer';
import { ITransformerRequireStatementCollection } from './interfaces/ITransformerRequireStatements';
import { ImportType } from './interfaces/ImportType';
import { parseTypeScript } from './parser';
import { ISerializableTransformationContext } from './transformer';

function cleanupForTest(node) {
  delete node.loc;
  delete node.raw;
  delete node.range;
}
export function initCommonTransform(props: {
  code: string;
  compilerOptions?: ICompilerOptions;
  jsx?: boolean;
  props?: ISerializableTransformationContext;
  transformers: Array<ITransformer>;
}) {
  const requireStatementCollection: ITransformerRequireStatementCollection = [];
  function onRequireCallExpression(importType: ImportType, statement: ASTNode) {
    // making sure we have haven't emitted the same property twice
    if (!statement['emitted']) {
      Object.defineProperty(statement, 'emitted', { enumerable: false, value: true });
      cleanupForTest(statement.arguments[0]);
      cleanupForTest(statement);
      cleanupForTest(statement.callee);

      requireStatementCollection.push({ importType, statement });
    }
  }

  const ast = parseTypeScript(props.code, { jsx: props.jsx });
  const userProps: ISerializableTransformationContext = props.props || {};
  userProps.compilerOptions = props.compilerOptions || {};
  const visitorProps: ITransformerCommon = { onRequireCallExpression, transformationContext: userProps };

  const tranformers = [];
  for (const t of props.transformers) {
    if (t.commonVisitors) {
      tranformers.push(t.commonVisitors(visitorProps));
    }
  }

  transformModule({ root: ast, transformers: tranformers });

  const res = generate(ast, {});

  return { code: res, requireStatementCollection };
}
