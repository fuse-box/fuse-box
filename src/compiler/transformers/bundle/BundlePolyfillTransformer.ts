import * as path from 'path';
import { ASTNode } from '../../interfaces/AST';
import { ITransformerSharedOptions } from '../../interfaces/ITransformerSharedOptions';
import { GlobalContext } from '../../program/GlobalContext';
import { createRequireStatement } from '../../Visitor/helpers';
import { IVisit, IVisitorMod } from '../../Visitor/Visitor';
import { ImportType } from '../../interfaces/ImportType';
import { ITarget } from '../../../config/PrivateConfig';

export interface IBundleEssentialProps {
  target?: ITarget;
  env?: { [key: string]: any };
  isBrowser?: boolean;
  isServer?: boolean;
  moduleFileName?: string;
}

export type IBundleEssntialTransformerOptions = IBundleEssentialProps & ITransformerSharedOptions;

export const PolyfillEssentialConfig = {
  stream: 'stream',
  buffer: 'buffer',
  Buffer: 'buffer',
  http: 'http',
  https: 'https',
};

export function BundlePolyfillTransformer(options: IBundleEssntialTransformerOptions) {
  const RequireStatementsInserted = {};
  const VariablesInserted = {};
  options = options || {};
  return (visit: IVisit): IVisitorMod => {
    const { node } = visit;

    if (visit.isLocalIdentifier) {
      const globalContext = visit.globalContext as GlobalContext;
      const locals = visit.scope && visit.scope.locals ? visit.scope.locals : {};
      const name = node.name;

      if (locals[name] || globalContext.hoisted[name]) {
        return;
      }

      switch (name) {
        case '__dirname':
          return { replaceWith: { type: 'Literal', value: options.moduleFileName } };
        case '__filename':
          return { replaceWith: { type: 'Literal', value: path.dirname(options.moduleFileName) } };
      }

      if (options.target !== 'browser') return;

      /**
       * *********************************************************************************
       * Polyfills stream buffer e.t.c
       */
      if (PolyfillEssentialConfig[name]) {
        if (VariablesInserted[name]) return;
        const statements: Array<ASTNode> = [];
        const moduleName = PolyfillEssentialConfig[name];
        if (!RequireStatementsInserted[moduleName]) {
          RequireStatementsInserted[moduleName] = 1;
          const statement = createRequireStatement(moduleName, name);
          if (options.onRequireCallExpression)
            options.onRequireCallExpression(ImportType.REQUIRE, statement.reqStatement);
          statements.push(statement.statement);
        }
        if (name !== moduleName) {
          statements.push({
            type: 'VariableDeclaration',
            kind: 'var',
            declarations: [
              {
                type: 'VariableDeclarator',
                init: {
                  type: 'Identifier',
                  name: moduleName,
                },
                id: {
                  type: 'Identifier',
                  name: name,
                },
              },
            ],
          });
        }
        if (statements.length) {
          return { prependToBody: statements };
        }
      }
    }
    return;
  };
}
