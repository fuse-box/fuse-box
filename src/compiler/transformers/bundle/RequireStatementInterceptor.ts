import { BUNDLE_RUNTIME_NAMES } from '../../../bundleRuntime/bundleRuntimeCore';
import { ISchema } from '../../core/nodeSchema';
import { ASTType } from '../../interfaces/AST';
import { ITransformer } from '../../interfaces/ITransformer';
import { ImportType } from '../../interfaces/ImportType';

export function RequireStatementInterceptor(): ITransformer {
  return {
    commonVisitors: props => {
      return {
        onEach: (schema: ISchema) => {
          const { getLocal, localIdentifier, node, parent, replace } = schema;

          if (localIdentifier) {
            if (node.name === 'require' && parent.type !== ASTType.CallExpression) {
              if (getLocal(node.name)) return;
              return replace({ name: BUNDLE_RUNTIME_NAMES.ARG_REQUIRE_FUNCTION, type: 'Identifier' });
            }
          }

          // handle typeof
          if (node.operator === 'typeof' && node.type === ASTType.UnaryExpression) {
            if (node.argument && node.argument.name) {
              const name = node.argument.name;
              // we must preserve local variable
              if (name === 'require') {
                if (getLocal(name)) return;
                return replace({ type: 'Literal', value: 'function' });
              }
            }
          }

          if (!props.onRequireCallExpression) return;

          if (node.type === 'CallExpression' && node.callee.name === 'require' && !node['emitted']) {
            if (!getLocal('require')) props.onRequireCallExpression(ImportType.REQUIRE, node);
          }
          return;
        },
      };
    },
  };
}
