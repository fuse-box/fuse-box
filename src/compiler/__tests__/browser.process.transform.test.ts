import { ITarget } from '../../config/ITarget';
import { initCommonTransform } from '../testUtils';
import { BrowserProcessTransformer } from '../transformers/bundle/BrowserProcessTransformer';
import { BundleFastConditionUnwrapper } from '../transformers/bundle/BundleFastConditionTransformer';
import { RequireStatementInterceptor } from '../transformers/bundle/RequireStatementInterceptor';

describe('Browser process transformer test', () => {
  const testTranspile = (props: { NODE_ENV?: string; code: string; target?: ITarget }) => {
    return initCommonTransform({
      code: props.code,
      compilerOptions: {
        buildTarget: props.target || 'browser',
        processEnv: { NODE_ENV: props.NODE_ENV || 'development' },
      },
      transformers: [BrowserProcessTransformer(), BundleFastConditionUnwrapper(), RequireStatementInterceptor()],
    });
  };
  it('should unwrap dev', () => {
    const res = testTranspile({
      code: `
      if ( process.env.NODE_ENV === "development"){
        console.log("dev")
      }
    `,
    });

    expect(res.code).toMatchSnapshot();
  });

  it('should unwrap dev with alternate', () => {
    const res = testTranspile({
      NODE_ENV: 'production',
      code: `
      if ( process.env.NODE_ENV === "development"){
        console.log("dev")
      } else {
        console.log("this is dev")
      }
    `,
    });
    expect(res.code).toMatchSnapshot();
  });

  it('should not emit require statement', () => {
    const res = testTranspile({
      NODE_ENV: 'production',
      code: `
      if ( process.env.NODE_ENV === "development"){
        require("./dev")
      } else {
        console.log("this is dev")
      }
    `,
    });

    expect(res.requireStatementCollection).toEqual([]);
  });

  it('should not touch other statements', () => {
    const res = testTranspile({
      NODE_ENV: 'production',
      code: `
      if ( a.b ) {
         console.log(1)
      } else {
        console.log(2)
      }
    `,
    });
    expect(res.code).toMatchSnapshot();
  });
});
