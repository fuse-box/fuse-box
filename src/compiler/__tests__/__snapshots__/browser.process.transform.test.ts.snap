// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Browser process transformer test should not touch other statements 1`] = `
"if (a.b) {
  console.log(1);
} else {
  console.log(2);
}
"
`;

exports[`Browser process transformer test should unwrap dev 1`] = `
"console.log(\\"dev\\");
"
`;

exports[`Browser process transformer test should unwrap dev with alternate 1`] = `
"console.log(\\"this is dev\\");
"
`;
